### PR TITLE
FEATURE: SD-759 Integrate <GeoChart> with uri/Visualization component

### PR DESCRIPTION
**Pull request by ptquang86**
---

[Check PR owner responsibilities](https://confluence.intgdc.com/display/Development/Code-reviews#Code-reviews-Ownerresponsibilities)

Supported PR commands:

Command | Description
--- | ---
`extended test - examples` | Live examples tests
`extended test - storybook` | Storybook screenshot tests

See more [options](https://confluence.intgdc.com/display/kbhr/How+to+work+with+git+and+Github#HowtoworkwithgitandGithub-Extendedchecks).

# PR Checklist
- [x] Verify code changes ([Checklist](https://confluence.intgdc.com/display/Development/Code-reviews+checklist), [Best practices](https://confluence.intgdc.com/display/Development/Code-reviews+best+practices))
- [x] [Verify pull-request formalities](https://confluence.intgdc.com/display/Development/Code-reviews)
- [x] Change was tested by using [gdc-dev-release](https://confluence.intgdc.com/display/~tomas.vojtasek/Private+NPM) in [gdc-analytical-designer](https://github.com/gooddata/gdc-analytical-designer) and [gdc-dashboards](https://github.com/gooddata/gdc-dashboards) (if applicable)
- [x] Migration guide (for major changes) is mentioned in [CHANGELOG.md](../blob/master/CHANGELOG.md).
- [x] Successful `extended test - examples`
- [x] Successful `extended test - storybook`
- [x] Checked yarn.lock consistency (no dep. duplicities especially Goodstrap)
- [x] Squash commits

# Related PRs
- gdc-analytical-designer: https://github.com/gooddata/gdc-analytical-designer/pull/2778
- gdc-dashboards: https://github.com/gooddata/gdc-dashboards/pull/2544

# Related Jira tasks
- SD-759: https://jira.intgdc.com/browse/SD-759

### CODE REVIEW

**Comment by ptquang86** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5910/ : SUCCESS in 5m 55s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1514/ : SUCCESS in 4s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3044/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1515/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2443/ : UNSTABLE in 17m 25s

**Comment by ptquang86** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5911/ : SUCCESS in 7m 22s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1516/ : SUCCESS in 5s

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1517/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2444/ : FAILURE in 1h 00m 21s

**Comment by ptquang86** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1520/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2446/ : FAILURE in 1h 00m 21s

**Comment by ptquang86** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5914/ : FAILURE in 7m 53s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1522/ : SUCCESS in 7s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3048/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1523/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2447/ : UNSTABLE in 15m 02s

**Comment by ptquang86** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5919/ : FAILURE in 1m 42s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1530/ : SUCCESS in 4s

**Comment by gdgate** [comment]
Build in pipeline check-extended aborted.

**Comment by ptquang86** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5920/ : SUCCESS in 4m 56s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1532/ : SUCCESS in 4s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3054/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1533/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2452/ : UNSTABLE in 14m 46s

**Comment by ptquang86** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5921/ : SUCCESS in 5m 39s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1535/ : SUCCESS in 4s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3056/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1536/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2454/ : SUCCESS in 14m 58s

**Comment by ptquang86** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-1361.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1537/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2455/ : SUCCESS in 27m 43s

**Comment by naibinh** [comment]
LGTM

File: src/helpers/MdObjectHelper.ts [hunk 1]
```diff
@@ -56,6 +66,77 @@ function buildArithmeticMeasureTitleProps(
...
+    );
+    const segmentByBucket = contentBuckets.find(
+        (bucket: VisualizationObject.IBucket): boolean => bucket.localIdentifier === SEGMENT_BY,
+    );
+    const sizeBucket = contentBuckets.find(
+        (bucket: VisualizationObject.IBucket): boolean => bucket.localIdentifier === SIZE,
+    );
+    const tooltipTextBucket = contentBuckets.find(
+        (bucket: VisualizationObject.IBucket): boolean => bucket.localIdentifier === TOOLTIP_TEXT,
+    );
```
[the remark this answers is not in the document]
Comment (ptquang86) [reply src/helpers/MdObjectHelper.ts]: good catching.
We already had a function for it named `findBucketByLocalIdentifier`.
Thanks

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5940/ : SUCCESS in 5m 12s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1552/ : SUCCESS in 4s

File: src/helpers/MdObjectHelper.ts [hunk 2]
```diff
@@ -56,6 +67,64 @@ function buildArithmeticMeasureTitleProps(
...
+    const tooltipTextBucket = findBucketByLocalIdentifier(contentBuckets, TOOLTIP_TEXT);
+
+    const color: IGeoPushpinChartBucketProps["color"] = colorBucket && colorBucket.items[0];
+    const location: IGeoPushpinChartBucketProps["location"] =
+        locationBucket && (locationBucket.items[0] as VisualizationObject.IVisualizationAttribute);
+    const segmentBy: IGeoPushpinChartBucketProps["segmentBy"] =
+        segmentByBucket && (segmentByBucket.items[0] as VisualizationObject.IVisualizationAttribute);
+    const size: IGeoPushpinChartBucketProps["size"] = sizeBucket && sizeBucket.items[0];
+    const tooltipText: VisualizationInput.IAttribute =
+        tooltipTextBucket && (tooltipTextBucket.items[0] as VisualizationObject.IVisualizationAttribute);
```
Comment (lanhq) [inline src/helpers/MdObjectHelper.ts]:
```
interface IGeoPushpinChartBucketProps {
    location: VisualizationInput.IAttribute;
    size?: VisualizationInput.AttributeOrMeasure;
    color?: VisualizationInput.AttributeOrMeasure;
    segmentBy?: VisualizationInput.IAttribute;
    filters?: VisualizationInput.IFilter[];
    sortBy?: VisualizationInput.ISort[];
}
```
then
`const color: IGeoPushpinChartBucketProps[“color”] = colorBucket && colorBucket.items[0];`

why we don't use 
`const color: VisualizationInput.AttributeOrMeasure = colorBucket && colorBucket.items[0];`

File: src/helpers/MdObjectHelper.ts [hunk 3]
```diff
@@ -56,6 +67,64 @@ function buildArithmeticMeasureTitleProps(
...
+    // allow us to show the smaller pushpins on top of bigger pushpins
+    const sortBy: IGeoPushpinChartBucketProps["sortBy"] = [];
+
+    const afmWithoutMergedFilters = DataLayer.toAfmResultSpec(content).afm;
+    afmWithoutMergedFilters.filters = afmWithoutMergedFilters.filters || [];
+
+    const afm = mergeFiltersToAfm(afmWithoutMergedFilters, filtersFromProps);
+
+    // Filter out expression filters which are not supported in bucket interface
+    const filters: IGeoPushpinChartBucketProps["filters"] = (afm.filters || []).filter(
```
Comment (lanhq) [inline src/helpers/MdObjectHelper.ts]: same here
`IGeoPushpinChartBucketProps["filters"]` => `VisualizationInput.IFilter[]`

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5941/ : SUCCESS in 6m 01s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1553/ : SUCCESS in 6s

**Comment by lanhq** [comment]
LGTM

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5944/ : SUCCESS in 15m 44s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1556/ : SUCCESS in 6s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5948/ : SUCCESS in 5m 27s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1561/ : SUCCESS in 5s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5955/ : SUCCESS in 7m 19s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1572/ : SUCCESS in 7s

**Comment by ptquang86** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5959/ : SUCCESS in 8m 48s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1577/ : SUCCESS in 5s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3075/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1578/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2467/ : SUCCESS in 17m 56s

**Comment by ptquang86** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-1361.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1581/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2469/ : FAILURE in 26m 28s

**Comment by ptquang86** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-1361.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1583/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2471/ : FAILURE in 13m 10s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by ptquang86** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-1361.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5963/ : SUCCESS in 9m 00s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1585/ : SUCCESS in 7s

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1586/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2472/ : SUCCESS in 30m 01s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5970/ : SUCCESS in 19m 24s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1598/ : SUCCESS in 8s

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3081/ : SUCCESS in 19m 12s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5973/ : SUCCESS in 15m 43s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/1434/ : SUCCESS in 23s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-examples-zuul-docker/1996/ : SUCCESS in 31m 28s

**Pull request merged**